### PR TITLE
Okina - RAJA (seq, omp, cuda) backend [artv3/okina-raja]

### PR DESCRIPTION
**Pull request by artv3**
Hi all, I've added a RAJA backend to Okina. Example 1 can be run with either the CPU or GPU using RAJA. We can extend the work here to enable an OpenMP RAJA backend as well. Currently I am relying on the existing memory manager to move data to and from the GPU.

### CODE REVIEW

File: examples/ex1.cpp [hunk 1]
```diff
@@ -53,8 +53,9 @@ int main(int argc, char *argv[])
    const char *mesh_file = "../data/star.mesh";
    int order = 1;
    bool static_cond = false;
-   bool pa = false;
+   bool pa = true;
```
Comment (jdahm) [inline examples/ex1.cpp]: Should switch this back to `false` before merging to preserve options

File: general/okina.hpp [hunk 2]
```diff
@@ -36,15 +40,31 @@
...
    {
-      for (size_t k=0; k<N; k+=1) { h_body(k); }
+#ifdef MFEM_USE_RAJA
+     printf("RAJA enabled! \n");
+     return RAJA::forall<RAJA::cuda_exec<BLOCKS>>(RAJA::RangeSegment(0,N), d_body);
+#else
+     MFEM_ABORT("RAJA requested for MFEM but RAJA is not enabled!");
+#endif
+
+  }else if (gpu) {
```
Comment (jdahm) [inline general/okina.hpp]: There are some minor formatting issues here through line 65. This should get taken care of with a `make style`.

File: general/okina.hpp [hunk 3]
```diff
@@ -36,15 +40,31 @@
 template <size_t BLOCKS, typename DBODY, typename HBODY>
 void wrap(const size_t N, DBODY &&d_body, HBODY &&h_body)
 {
-   const bool gpu = mfem::config::usingGpu();
-   if (gpu)
-   {
-      return cuWrap<BLOCKS>(N,d_body);
-   }
-   else
+   const bool gpu = mfem::config::usingGpu();   //
```
Comment (jdahm) [inline general/okina.hpp]: Empty comments here and on line 44.

File: general/okina.hpp [hunk 4]
```diff
@@ -36,15 +40,31 @@
...
+   const bool raja = mfem::config::usingRaja(); //
+   if(gpu && raja)
    {
-      for (size_t k=0; k<N; k+=1) { h_body(k); }
+#ifdef MFEM_USE_RAJA
+     printf("RAJA enabled! \n");
```
Comment (jdahm) [inline general/okina.hpp]: Remove `printf()` call here before merging.

File: setupMFEM.sh [hunk 5]
```diff
@@ -0,0 +1,5 @@
+make config MFEM_CXX="nvcc" \
```
Comment (jdahm) [inline setupMFEM.sh]: This is certainly convenient, but should be incorporated into the MFEM makefiles (and maybe CMake? - @camierjs) before merging.

**Comment by jdahm** [comment]
Minor comments above - this is on the right track. It's good to see that this requires a relatively light touch to get to work with RAJA. Nice work!

File: general/okina.hpp (hunk 2 above)
Comment (artv3) [reply general/okina.hpp]: Thanks! I have to admit I'm not familiar with make style though....  Is it supported on the LLNL machines? If I just run make style I get the following: 
``$ make style
/bin/sh: astyle: command not found
No source files were changed.``

File: setupMFEM.sh (hunk 5 above)
Comment (artv3) [reply setupMFEM.sh]: Definitely! It was a cheat script for me. I wasn't sure on how things in the build system should be modified, but I can made adjustments accordingly.

File: general/okina.hpp (hunk 2 above)
Comment (jdahm) [reply general/okina.hpp]: In order to use `make style`, you need `astyle` installed and in your shell's `PATH` environment variable. Unfortunately, I am not familiar with the LLNL desktops. If it's Red Hat Enterprise Linux, then you may be able to install it with the yum package manager...

File: general/config.cpp [hunk 6]
```diff
@@ -35,6 +35,26 @@ void config::CudaDeviceSetup(const int device)
...
+void config::RajaDeviceSetup(const int device)
+{
+#ifdef __NVCC__
+   cudaGetDeviceCount(&ngpu);
+   MFEM_ASSERT(ngpu>0, "No CUDA device found!");
+   cuInit(0);
+   dev = device;
+   cuDeviceGet(&cuDevice,dev);
+   cuCtxCreate(&cuContext, CU_CTX_SCHED_AUTO, cuDevice);
+   cuStream = new CUstream;
```
Comment (jdahm) [inline general/config.cpp]: Is cuStream ever deleted?

Comment (artv3) [reply general/config.cpp]: Good catch, I just added a destructor for the config class.

File: general/okina.hpp (hunk 2 above)
Comment (artv3) [reply general/okina.hpp]: Cool thanks!

File: general/okina.hpp [hunk 7]
```diff
@@ -36,15 +40,37 @@
...
+#else
+      MFEM_ABORT("RAJA requested for MFEM but RAJA is not enabled!");
+#endif
+
+   }
    else
    {
       for (size_t k=0; k<N; k+=1) { h_body(k); }
    }
+
```
Comment (artv3) [inline general/okina.hpp]: @jdahm @camierjs , RAJA also supports OpenMP and TBB backends. We could always add additional if statements, but wasn't sure if there was a better way. What do you guys think?

**Pull request merged**